### PR TITLE
txn:improvement for point get by unique key or pk 

### PR DESCRIPTION
**Pull request by AndreMouche**
This PR implement improvement for point get by unique key or pk with tikv:
For A point get by unique key or pk, use uint64::max  as start_ts instead of getting from pd. And for this case, Tikv would return the latest committed version if lock exist and key is the lock's primary.
@disksing @coocood PTAL

### CODE REVIEW

**Comment by siddontang** [comment]
PTAL @hhkbp2

File: kv/kv.go [hunk 1]
```diff
@@ -176,6 +176,8 @@ type Driver interface {
 type Storage interface {
 	// Begin transaction
 	Begin() (Transaction, error)
+	// Begin transaction with startTs
+	BeginWithStartTs(startTs uint64) (Transaction, error)
```
Comment (disksing) [inline kv/kv.go]: How about create a _Snapshot_ instead? It should be readonly anyway.

File: store/tikv/kv.go [hunk 2]
```diff
@@ -145,6 +145,16 @@ func (s *tikvStore) Begin() (kv.Transaction, error) {
 	return txn, nil
 }
 
+// BeginPointGetByPkOrUniqueKey begins a transaction for point get by pk or unique key
+func (s *tikvStore) BeginWithStartTs(startTs uint64) (kv.Transaction, error) {
```
Comment (disksing) [inline store/tikv/kv.go]: Any mechanism to make sure that only 1 point get call can be made on this txn?

File: executor/adapter.go [hunk 3]
```diff
@@ -24,6 +24,7 @@ import (
 	"github.com/pingcap/tidb/infoschema"
 	"github.com/pingcap/tidb/model"
 	"github.com/pingcap/tidb/plan"
+	"math"
```
Comment (hanfei1991) [inline executor/adapter.go]: move it to line 18

File: kv/kv.go (hunk 1 above)
Comment (coocood) [inline kv/kv.go]: Our implementation requires that every request have an associated transaction.

File: context/context.go [hunk 4]
```diff
@@ -47,6 +47,9 @@ type Context interface {
 	// ActivePendingTxn receives the pending transaction from the transaction channel.
 	// It should be called right before we builds an executor.
 	ActivePendingTxn() error
+	// ActivePendingTxnWithStartTs receives the pending transaction with start_ts.
+	// It should be called right before we builds an executor.
+	ActivePendingTxnWithStartTs(startTs uint64) error
```
Comment (coocood) [inline context/context.go]: `TS` instead of `Ts`.

File: executor/adapter.go [hunk 5]
```diff
@@ -80,6 +81,26 @@ func (a *statement) OriginText() string {
...
+	checkPlan := a.plan
+	if proj, ok := checkPlan.(*plan.Projection); ok {
+		if len(proj.Children()) != 1 {
+			return false
+		}
+		checkPlan = proj.Children()[0]
+	}
+
+	if indexScan, ok := checkPlan.(*plan.PhysicalIndexScan); ok {
+		return len(indexScan.Ranges) == 1 && indexScan.Ranges[0].IsPoint(ctx.GetSessionVars().StmtCtx)
```
Comment (coocood) [inline executor/adapter.go]: It's not enough to check index range is point.
Need to check index is unique and low value count is the same index columns count.

**Comment by coocood** [comment]
`executor/prepared.go:237` can apply this optimization too.

File: kv/kv.go (hunk 1 above)
Comment (disksing) [inline kv/kv.go]: I thought the history read feature is based on creating a snapshot at some user defined TS. Can this case handle it in a similar way?

File: executor/prepared.go [hunk 6]
```diff
@@ -24,6 +24,7 @@ import (
 	"github.com/pingcap/tidb/parser"
 	"github.com/pingcap/tidb/plan"
 	"github.com/pingcap/tidb/util/sqlexec"
+	"math"
```
Comment (zimulala) [inline executor/prepared.go]: Move this line to line 17.

File: store/tikv/txn.go [hunk 7]
```diff
@@ -57,6 +57,18 @@ func newTiKVTxn(store *tikvStore) (*tikvTxn, error) {
 	}, nil
 }
 
+// newTikvTxnWithStartTS creates a txn with startTS
```
Comment (hanfei1991) [inline store/tikv/txn.go]: add `.`

File: store/tikv/kv.go [hunk 8]
```diff
@@ -145,6 +145,16 @@ func (s *tikvStore) Begin() (kv.Transaction, error) {
 	return txn, nil
 }
 
+// BeginWithStartTS begins a transaction with startTS
```
Comment (hanfei1991) [inline store/tikv/kv.go]: add `.`

File: store/localstore/kv.go [hunk 9]
```diff
@@ -371,6 +371,11 @@ func (s *dbStore) Begin() (kv.Transaction, error) {
 	return newTxn(s, beginVer), nil
 }
 
+// BeginWithStartTS begins transaction with startTS
```
Comment (hanfei1991) [inline store/localstore/kv.go]: add `.`

File: session.go [hunk 10]
```diff
@@ -908,3 +908,25 @@ func (s *session) ActivePendingTxn() error {
...
+	var err error
+	s.txn, err = s.store.BeginWithStartTS(startTS)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	err = s.loadCommonGlobalVariablesIfNeeded()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	log.Debug("InitTxnWithStartTS with ", s.txn.StartTS())
```
Comment (hanfei1991) [inline session.go]: remove this line.

File: plan/physical_plans.go [hunk 11]
```diff
@@ -500,6 +501,14 @@ func (p *PhysicalIndexScan) MarshalJSON() ([]byte, error) {
 	return buffer.Bytes(), nil
 }
 
+// IsPointGetByUniqueKey checks whether is a point get by unique key
```
Comment (hanfei1991) [inline plan/physical_plans.go]: add `.`

File: kv/mock.go [hunk 12]
```diff
@@ -103,8 +103,13 @@ func (s *mockStorage) Begin() (Transaction, error) {
 		valid: true,
 	}
 	return tx, nil
+}
 
+// BeginWithStartTS begins a transaction with startTS
```
Comment (hanfei1991) [inline kv/mock.go]: add `.`

File: executor/adapter.go [hunk 13]
```diff
@@ -167,3 +175,39 @@ func (a *statement) logSlowQuery() {
...
+	if !ctx.GetSessionVars().IsAutocommit() {
+		return false
+	}
+
+	// check txn
+	if ctx.Txn() != nil {
+		return false
+	}
+
+	// check plan
```
Comment (hanfei1991) [inline executor/adapter.go]: You can write a loop to check if this plan is a chain.

**Comment by disksing** [comment]
Any benchmark result?

**Comment by AndreMouche** [comment]
Point Read TPS  improved 3%  under conditions of 32 thread,1 `tikv`,1 `tidb`,1 `pd` @disksing

**Comment by shenli** [comment]
@AndreMouche Please address the comments from hanfei

File: executor/adapter.go (hunk 13 above)
Comment (coocood) [inline executor/adapter.go]: The chance of chained projection is very low.
I think just check one level is enough.

**Comment by ngaut** [comment]
PTAL

**Comment by AndreMouche** [comment]
@zimulala @coocood @hanfei1991 @hanfei19910905 PTAL

**Comment by coocood** [comment]
LGTM

**Pull request merged**